designs/0032-stan-output-formats.md

```diff
@@ -0,0 +1,295 @@
+straightforward representation of the posterior sample one row per draw,
+one column per output from the Stan program.
+Over time, the Stan CSV file has come to be an amalgam of information about the inference engine configuration,
+the algorithm state, and the algorithm outputs.
+The formats of the per-inference algorithm outputs are described in the
+[appendix of the CmdStan User's Guide](https://mc-stan.org/docs/2_29/cmdstan-guide/stan_csv.html)
+
+### Output mechanism:  callback writers
+
+Outputs are handled by objects of base class `stan::callbacks::writer`
+which is constructed from an output stream.
+The use of writer callbacks keeps the inference algorithms output-format agnostic.
+Through function call operator overloading, the writer
+can handle the following data structures:
+
++ a vector of strings (names)
++ a vector of doubles (data)
++ a string (message)
++ nothing - blank input
+
+The `stan::services` layer provides the set of functions which wrap the inference algorithms.
+To do inference given a model, data, and config,
+the Stan interfaces first instantiate the Stan model object and the writer callback,
+which are passed in to the appropriate `stan::services` function.
+The sampling and variational services provide arguments
+for both a output and an auxiliary output writers (`diagnostic_writer`)
+while the optimization and standalone generated quantities services only provide a single output writer.
+In CmdStan, the callback instance is a `stan::callbacks::stream_writer` object
+which formats data as CSV comments, header row, and data rows.
+The PyStan and RStan interfaces provide R and Python appropriate writer objects.
+
+### Stan program outputs
+
+The Stan model class `write_array` method returns a vector of doubles for all
+parameter, transformed parameter, and generated quantities (non-local) variables,
+in order of declaration in the Stan program.
+
+The model object's member functions `constrained_param_names`, `unconstrained_param_names`, `get_constrained_sizedtypes`, `get_unconstrained_sizedtypes`
+provide name, type, and shape information.
+The reason that the model reports on both the constrained and unconstrained parameters is the Stan parameter types
+simplex, correlation, covariance, and Cholesky factors have different shapes on the constrained, unconstrained scale.
+For example, a simplex parameter is an N vector whose elements sum to one on the constrained scale which means that it has N-1 free parameters
+when computing on the unconstrained scale.
+The Stan model parameter inputs and outputs are on the *constrained* scale, therefore the number of parameter values reported by the `write_array` method
+is reported by the `get_constrained_sizedtypes` method.
+The sampler computes on the *unconstrained* scale, therefore the size of the metric (inverse mass matrix) is reported by the `get_unconstrained_sizedtypes` method.
+
+### Inference algorithm outputs
```

Comment:
Unclear if this is describing the current status or what you're proposing. If the current, will this structure be maintained or changed (and how?)